### ISSUE TITLE
New format for dataset metadata

### ISSUE THREAD

**Issue by kohya-ss** [kohya-ss/sd-scripts#1752]
## Format
The file is in JSON format.

```
{
  "format_version": "1.0", 
  "latent_cache_dir": "/path/to/latent_cache_dir",  // optional, for all images in this .json
  "images": {
    "/full/path/to/image.png": { // image file
      "caption": ["caption1 for the image", "another caption for the image", ... ], 
      "tags": ["tag1, tag2, tag3", "another tag1, ...", ...], 
      "caption_weights": [0.5, 0.1, ...], // optional, the sum does not have to be 1
      "tags_weights": [0.2,1.5, ...]
      "image_size": [1024, 768], 
      "latent_cache_dir": "/path/to/latent_cache_dir"  // optional, for this image
    }, 
    "/full/path/to/archive.tar////inner/path/of/image.png": { // image in archive file
      "caption": ["caption for the image"], 
      "tags": ["tag1, tag2, tag3"], 
      "image_size": [1024, 768], 
      "latent_cache_dir": "/path/to/latent_cache_dir"  // optional, for this image
    }, 
    ...
  }
}
```

- Contains image files and image files in archives.
- When specifying a file within an archive, the archive filename and the innter path in the archive are separated by `////`. A key containing `////` indicates a file within an archive.
- `latent_cache_dir` is optional. If omitted, the image file directory will be used.
- `latent_cache_dir` cannot be omitted if an archive file is specified.
- Either `caption` or `tags` can be omitted.
- Multiple captions and tags can be specified in a list.
- Each image may have both `caption` and `tags`, but how they are handled depends on the specifications of the training script. Currently they are simply concatenated.

Initially, `.tar` is the only archive type expected, but support may expand in the future. EDIT: `.zip` is also supported.

Please feel free to share your thoughts and suggestions on this specification.

EDIT: Add `caption_weights` and `tags_weights` for the probabilities.

**Comment by ChristianMayer** [comment]
I expect `.zip` be more common than `.tar` for image archives - even on Linux.

Would providing `caption` and `tag` be a way to use one for T5 and the other one for CLIP for models that have both (SD3, Flux)?

**Comment by shirayu** [comment]
> I expect .zip be more common than .tar for image archives - even on Linux.

I agree with that as well. 
`.zip` is easy to extract some files without decompressing the entire file, while `.tar` cannot do the same thing.
Additionally, `zip` can also be compressed (uncompressed is also possible)

**Comment by alfredplpl** [comment]
I agree with @ChristianMayer .
And, I also suggest .gz file.
It is easier to write the code like this:
```python
import json
import gzip

with gzip.open(filename, 'r') as f:
    for line in f:
        obj = json.loads(line)
        do_something(obj)
```
See also: https://fnwiya.hatenablog.com/entry/2016/01/07/210000

**Comment by sdbds** [comment]
Describe how I manage datasets locally
1、First create a lance file from pyarrow use table:
```
            pa.field("ImageURI", pa.string()),
            pa.field("extension", pa.string()),
            pa.field("hash", pa.string()),
            pa.field("size", pa.int64()),
            pa.field("width", pa.int32()),
            pa.field("height", pa.int32()),
            pa.field("binary", pa.binary()),
            pa.field("captions", pa.list_(pa.string(), -1)),
```

2、loading from website / huggingface /local filesystem

there will be version 1 like this
```
                    "ImageURI" :s3://example/image3.png ,
                    "extension":png,
                    "hash":none,
                    "size":none,
                    "width":none,
                    "height":none,
                    "binary":none,
                    "captions":none,
```

3、Then open image to get metadata as version 2(if from internet it will download then save in binary)

```
                    "ImageURI" :s3://example/image3.png ,
                    "extension":png,
                    "hash":0d9350f0d132236f201e121aa5a65bff5a0cfe13ae08fa08a2e92b035047c97f,
                    "size":123124123123,
                    "width":1024,
                    "height":1024,
                    "binary":b'\x89PNG\r\n\x1a\n\x00\x00\x00\rIHDR\x00\x00\x00...',
                    "captions":none,
```

4、unzip binary export to image dir then update imageURI to local  (if localsystem dont exist image)

5、enable ARB bucket in here for reduce image size.

6、captions with image then save as version 3

```
                    "ImageURI" :D:\\example\\image3.png ,
                    "extension":png,
                    "hash":0d9350f0d132236f201e121aa5a65bff5a0cfe13ae08fa08a2e92b035047c97f,
                    "size":123124123123,
                    "width":1024,
                    "height":1024,
                    "binary":b'\x89PNG\r\n\x1a\n\x00\x00\x00\rIHDR\x00\x00\x00...',
                    "captions":[tag,caption1,caption2],
```

7、(optional) delete binary in lance for reducing file size

8、cache latent files then save as version 4

```
                    "ImageURI" :D:\\example\\image3_sd3.npz ,
                    "extension":npz,
                    "hash":0d9350f0d132236f201e121aa5a65bff5a0cfe13ae08fa08a2e92b035047c97f,
                    "size":234234234,
                    "width":1024,
                    "height":1024,
                    "binary":b'\',
                    "captions":[tag,caption1,caption2],
```

9、(optional) delete origin Image in lance for reducing file size

Finally we get image files、captions and dataset metadata
And we can switch to origin state in any version.
such as before arb、from arb to caption、before cache latent...

**Comment by ChristianMayer** [comment]
> > Would providing `caption` and `tag` be a way to use one for T5 and the other one for CLIP for models that have both (SD3, Flux)?
> 
> This is certainly an interesting approach, but I don't think FLUX.1 or SD3 have been pre-trained in this way, so I would like to consider this in the future.

I also think so. But people are using it already with split captions so I wouldn't be surprised when people want to train it in this way as well.

**Comment by sdbds** [comment]
> > > Would providing `caption` and `tag` be a way to use one for T5 and the other one for CLIP for models that have both (SD3, Flux)?
> > 
> > 
> > This is certainly an interesting approach, but I don't think FLUX.1 or SD3 have been pre-trained in this way, so I would like to consider this in the future.
> 
> I also think so. But people are using it already with split captions so I wouldn't be surprised when people want to train it in this way as well.

Many people have mentioned this method, but I don't think it's reliable.
Since SDXL there have been attempts to train CLIP G and CLIP L separately, where one is responsible for the details and the other for the whole...
But from the test results, they work together, and training them separately would actually be similar to the effect of dropout separately, which would lead to a lot of randomness in the results unless a lot of training is used to pile up the generalizability.
